### PR TITLE
GSoC - Doctor Notes Enhancement 

### PR DESCRIPTION
**Pull request by UdaySagar-Git**
This PR is created to track my progress on the GSoC project (#7159) and to merge all PRs related to Doctor notes.

Backend RR: https://github.com/coronasafe/care/pull/2295

https://github.com/user-attachments/assets/193d2775-e83b-417c-af9a-16fdf67c19bd

- [x] Reply feature in doctore notes
- [x] Profile image support for a user
- [x] Markdown Support 
--------------------
File Upload Option
- [x] support for camera captures
- [x] support for audio recording
- [x] support for uploading files
- [x] file preview support foNr uploaded files
----
User Tagging Mechanism
- [x] support for tagging a user (mentions) 
- [x] Filtering of users and support for selecting users with keyboard arrows
- [x] sending notifications upon mention
- [x] hover card to show user details 
---
- [x] Show Uploaded files under patient consultation page
---

## Release Notes

- **New Features**
  - Introduced a new notification type for "Patient Note Mentioned."
  - Added support for new file types in the File Preview Dialog, including MP3.
  - Enhanced Markdown rendering with user mention functionality and custom link handling.
  - Introduced a Mentions Dropdown for user selection based on input.
  - Launched a Rich Text Editor for creating and formatting notes.
  - Added a Discussion Notes Files tab in the Consultation Details component.
  - Implemented detailed views for patient notes and replies.
  - Introduced a new component for managing and displaying discussion notes files.

- **Bug Fixes**
  - Updated URL handling for notifications to ensure proper routing.

- **Documentation**
  - Added new localization keys for the consultation tab.

- **Chores**
  - Improved type safety and error handling across various components.

### CODE REVIEW

**Comment by UdaySagar-Git** [comment]
@rithviknishad  can you please remove stale label

File: src/Components/Common/MarkdownPreview.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    if (facilityUsers?.results) {
+      const newCache: Record<string, UserModel> = {};
+      facilityUsers.results.forEach((user) => {
+        if (user.username) {
+          newCache[user.username] = user as UserModel;
+        }
+      });
+      setUserCache(newCache);
+    }
+  }, [facilityUsers]);
```
Comment (rithviknishad) [inline src/Components/Common/MarkdownPreview.tsx]: @UdaySagar-Git the user cache is for showing user information for mentioned users right?

If going this way, we could consider having the cache state at app level instead in the component level.

But it'd be better for the notes the serializer itself to serialize the user details from backend, maybe as a separate attribute like how GitHub's APIs are. The note object would look something like this:

```
body: string
mentionedUsers: Array<UserMention>
```

File: src/Components/Common/MarkdownPreview.tsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      </h3>
+      <p className="text-xs text-gray-500">@{user.username}</p>
+      <p className="text-xs text-gray-500">{user.user_type}</p>
+    </div>
+  </div>
+);
+
+const MarkdownPreview: React.FC<{ markdown: string }> = ({ markdown }) => {
+  const [userCache, setUserCache] = useState<Record<string, UserModel>>({});
+  const facilityId = "81092ced-8720-44cb-b4c5-3f0ad0540153";
```
Comment (rithviknishad) [inline src/Components/Common/MarkdownPreview.tsx]: I assume this was hardcoded for testing purposes. Shouldn't be there in final version.

File: src/Components/Common/MarkdownPreview.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+  const processedMarkdown = markdown
+    .replace(
+      /!\[mention_user\]\(user_id:(\d+), username:([^)]+)\)/g,
+      (_, userId, username) =>
+        `<a class="user-mention" href="/user/profile/${username}" data-user-id="${userId}" data-username="${username}">@${username}</a>`,
+    )
+    .replace(/~(.*?)~/g, (_, text) => `<del>${text}</del>`);
+
+  const CustomLink: React.FC<any> = (props) => {
```
Comment (rithviknishad) [inline src/Components/Common/MarkdownPreview.tsx]: `any` types are not allowed!

File: src/Components/Common/MentionDropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import React, { useEffect, useState } from "react";
+import routes from "../../Redux/api";
+import useQuery from "../../Utils/request/useQuery";
+
+const MentionsDropdown: React.FC<{
+  onSelect: (user: any) => void;
```
Comment (rithviknishad) [inline src/Components/Common/MentionDropdown.tsx]: `any` types are not allowed. Change all usages of `any` types introduced in this PR to instead have proper types.

File: src/Components/Patient/PatientNotes.tsx [hunk 5]
```diff
@@ -34,6 +35,9 @@ const PatientNotes = (props: PatientNotesProps) => {
   const [reload, setReload] = useState(false);
   const [facilityName, setFacilityName] = useState("");
   const [patientName, setPatientName] = useState("");
+  const [reply_to, setReplyTo] = useState<PatientNotesModel | undefined>(
+    undefined,
+  );
```
Comment (rithviknishad) [inline src/Components/Patient/PatientNotes.tsx]: Shorter equivalent for the same (with naming convention corrected)

```suggestion
  const [replyTo, setReplyTo] = useState<PatientNotesModel>();
```

File: src/Routers/routes/UserRoutes.tsx [hunk 6]
```diff
@@ -6,4 +7,11 @@ export default {
   "/users": () => <ManageUsers />,
   "/users/add": () => <UserAdd />,
   "/user/profile": () => <UserProfile />,
+  "/test": () => (
+    <RichTextEditor
+      initialMarkdown={""}
+      onChange={(value) => console.log(value)}
+      onAddNote={() => console.log("Add Note")}
+    />
+  ),
```
Comment (rithviknishad) [inline src/Routers/routes/UserRoutes.tsx]: I assume this'd be removed in the final version

**Review by rithviknishad — CHANGES_REQUESTED** [review]

File: src/Components/Common/MarkdownPreview.tsx (hunk 1 above)
Comment (UdaySagar-Git) [reply src/Components/Common/MarkdownPreview.tsx]: yes ! will change that to get details of mentioned users directly while fetching notes, will take it up after midterm

File: src/Components/Common/MarkdownPreview.tsx (hunk 2 above)
Comment (UdaySagar-Git) [reply src/Components/Common/MarkdownPreview.tsx:29]: will change

File: src/Components/Common/MarkdownPreview.tsx (hunk 3 above)
Comment (UdaySagar-Git) [reply src/Components/Common/MarkdownPreview.tsx]: got it

File: src/Components/Common/MentionDropdown.tsx (hunk 4 above)
Comment (UdaySagar-Git) [reply src/Components/Common/MentionDropdown.tsx]: got it

File: src/Routers/routes/UserRoutes.tsx (hunk 6 above)
Comment (UdaySagar-Git) [reply src/Routers/routes/UserRoutes.tsx]: yes

**Comment by nihal467** [comment]
@UdaySagar-Git 

- [ ] the cypress is failing, fix it



image

- [ ] the attached image or audio files are not been attached to the chat.


image

- [ ] when a invalid file type is upload, the chat space is not restore the state, it is still showing 100% uploaded all the time
- [ ] rather than showing invalid file type, mention more user friendly error message

image

- [ ] i can't remove applied bold or italic or any mark editor to the text, by again clicking on that particular icon

image

- [ ] when i apply a mark editor , the text is modified instantly but to see the highlight of any applied mark's i need to click outside the icons.

image
image
image

- [ ] for hyperlinking, add a pop-up similar to the gmail, where we can see the text we are hyperlinking and the link we want to redirect
- [ ] the hyperlinked text should be highlight as blue, not as black
- [ ] i try to hyperlink to google.com, but when i click on the link, i initially redirected to a page not found page, after that i was redirected to google.com, it should ideally redirect to the link
- [ ] when we click on the hyperlink, always have a browser behavior to open it on a new tab, so the user don't miss the chat

image
image
image

- [ ] try to inspire the tagging mechanism from slack, when we type, the text is visible to the user, and based on each character input, the list keep on filtering

image

- [ ] on hovering over the icon, add a tooltip to exact the full name of the icon lets say if user hover over B, the tooltip should say bold, similar for all icons

image

- [ ] what is the purpose of the above icon.

image

- [ ] when i upload a new profile photo, the success notification is shown in the platform, to view the upload new profile, i need to manually reload the page to refresh the state of the page

image
image


- [ ] the icons are clipping in and why are we always showing no replies in text. We could use any icon to represent replies if exist , and no icon if no replies exist


image

- [ ] keep the default view as the default, not the thread view

image

- [ ] if a user was tagged in the note, they should be send a notification stating that he/she has been mentioned in this note.

image

- [ ] when you click on the tagged user name, you are been taken to a page not found space,


**UI Suggestion** 

image

- [ ] the warning message for mike looks weirds always present in the UI, it is fine with file upload component, but in this space it doesn't look good, try to think of a better UI to place in her in the chat

image

- [ ] you are reusing the voice recording component here, it doesnt go with the UI , think of a better UI

image
image


- [ ] change the cancel text for reply comment to a icon and the send icons giving a feel it doesn't suit there, for example the slack editor looks good for the icon positioning

**Comment by khavinshankar** [comment]
@UdaySagar-Git fix all the issues suggested by @nihal467 (https://github.com/coronasafe/care_fe/pull/8016#issuecomment-2217740101) and push the code asap.

**Comment by UdaySagar-Git** [comment]
@khavinshankar sure! i will try to update by tommorrow

**Comment by nihal467** [comment]
@UdaySagar-Git once you fix all the issues, before requesting a re-review , attach a functionality video of the PR, to ensure that all the basic functionalities are working.

**Comment by UdaySagar-Git** [comment]
sure @nihal467

**Comment by UdaySagar-Git** [comment]
- [x]  the cypress is failing, fix it
- [x]  i can't remove applied bold or italic or any mark editor to the text, by again clicking on that particular icon
- [x]  when i apply a mark editor , the text is modified instantly but to see the highlight of any applied mark's i need to click outside the icons.
- [x]  for hyperlinking, add a pop-up similar to the gmail, where we can see the text we are hyperlinking and the link we want to redirect
- [x]  the hyperlinked text should be highlight as blue, not as black
- [x]  i try to hyperlink to google.com, but when i click on the link, i initially redirected to a page not found page, after that i was redirected to google.com, it should ideally redirect to the link
- [x]  when we click on the hyperlink, always have a browser behavior to open it on a new tab, so the user don't miss the chat
- [x]  on hovering over the icon, add a tooltip to exact the full name of the icon lets say if user hover over B, the tooltip should say bold, similar for all icons
- [x]  when i upload a new profile photo, the success notification is shown in the platform, to view the upload new profile, i need to manually reload the page to refresh the state of the page
- [x]  the icons are clipping in and why are we always showing no replies in text. We could use any icon to represent replies if exist , and no icon if no replies exist
- [x]  keep the default view as the default, not the thread view
- [x]  change the cancel text for reply comment to a icon and the send icons giving a feel it doesn't suit there, for example the slack editor looks good for the icon positioning
- [x]  when you click on the tagged user name, you are been taken to a page not found space,
---
Attachments and mentions enhancements are planned after mid term

- [ ]  the attached image or audio files are not been attached to the chat.
- [ ]  the warning message for mike looks weirds always present in the UI, it is fine with file upload component, but in this space it doesn't look good, try to think of a better UI to place in her in the chat
- [ ]  you are reusing the voice recording component here, it doesnt go with the UI , think of a better UI
- [ ]  when a invalid file type is upload, the chat space is not restore the state, it is still showing 100% uploaded all the time
- [ ]  rather than showing invalid file type, mention more user friendly error message
- [ ]  try to inspire the tagging mechanism from slack, when we type, the text is visible to the user, and based on each character input, the list keep on filtering
- [ ]  if a user was tagged in the note, they should be send a notification stating that he/she has been mentioned in this note.

**Comment by UdaySagar-Git** [comment]
https://github.com/coronasafe/care_fe/assets/111575806/28296e0a-7a9f-4ff3-93eb-74beb0d64ab3

File: src/Components/Common/MarkdownPreview.tsx (hunk 2 above)
Comment (UdaySagar-Git) [reply src/Components/Common/MarkdownPreview.tsx]: yupp

**Comment by nihal467** [comment]
@UdaySagar-Git 

**General Description**
---

- [ ] Modify the PR description properly to understand, what are covered specifically in this PR point by point so that the review will be easier, for example when we read:  File Upload Option (partially) in the description, we are not sure, which is completed and which are under development.

**Compared Review Based on Proposal till July 7th**
---

image
image
image
image


### Action Item till this date, based on contributors proposal
---
- [ ] Implementing Markdown with full functionality
- [ ] Adding support to embed, preview, and copy-paste images ( 1/3 only done )
- [ ] Implement the tagging mechanism of users 
- [ ] Implement the preview mode for hyperlinked links ( Missing in the PR )
- [ ] Introduce the file upload support in the doctor's notes (Partially only done)
- [ ] Make necessary changes to support bookmark in various categories ( Missing in the PR )
- [ ] Fully implementing support of file, audio, and camera capture (By July 21st)


### Test Report ( In Progress )
---

### Backend PR

- Backend test are failing

###  Markdown with full functionality

image
image
image

- the icons in the mark editor have multiple sizes, keep the icon size consistent.
- always use care-icon, I observed that you used a custom icon for the quote 

image
image
image
image


- Before starting to type the text, we should be able to select a text preference, such as bold, italic, or any other style, and keep it selected by default so that any text typed afterward follows that preference.
- When I select a particular part of a word and make it bold, the text becomes bold, but the bold icon does not turn green unless I move the focus away from the element.
- When I apply bold to a particular part of a word and continue typing, the bold preference is also being applied to the text I type afterward.
- When I type "How are you" and strike through the sentence, then remove it and start typing again from the beginning, the previously applied text style is being applied to the new text.

image
image
image


- When I hover over the hyperlinked text, the tooltip should display the URL to which we are being redirected.
- Currently, there is no option to edit an existing hyperlink that is already applied to the text. For example, as shown in the screenshot, the text is already hyperlinked, but when I select it again and click on the hyperlink option, it asks for a new link instead of allowing me to edit the existing one.
- when i hyperlink to an external link, the page is been redirected properly to the external link

image

- In the current pop-up, we only have option to add a link, so add an option to add text as well. This will allow users to type the text along with the link. Upon submitting the pop-up, the hyperlinked text should get autofilled in the chat box. This text box in the pop-up can also be used to show the selected text if the user has already selected any text to hyperlink from the chatbox.

image
image


- When i upload a single file, they are been uploaded twice
- when you try to upload a file, it's not being uploaded to the chat
- when the file name is larger, it's overflowing the card 

image

- When I click on the camera, the image gets auto-attached to the chat even before I click submit in the pop-up.

image

- enhancement to the tagging mechanism needs to be done as per the last comment

image

- the icons are still overlapping with the text in the mobile view

image

- When trying to view all replies to a message in the mobile view, an unwanted scrollbar appears.

image

- [ ] once we upload a user profile photo, you are refreshing the entire platform, don't do that, only refresh the component to show the uploaded photo, i think there is a provision in react to do that

**Comment by UdaySagar-Git** [comment]
@nihal467 updated the description, i will finish the mentioned tasks without any delays

> ### Action Item of the Week
> * [ ]  Implementing Markdown with full functionality
> * [ ]  Adding support to embed, preview, and copy-paste images
> * [ ]  Implement the tagging mechanism of users
> * [ ]  Implement the preview mode for hyperlinked links
> * [ ]  Introduce the file upload support in the doctor's notes
> * [ ]  Make necessary changes to support bookmark in various categories
> * [ ]  Fully implementing support of file, audio, and camera capture (By July 21st)

**Comment by khavinshankar** [comment]
@UdaySagar-Git any progress on this (https://github.com/coronasafe/care_fe/pull/8016#issuecomment-2222584799).

**Comment by UdaySagar-Git** [comment]
@nihal467  The following are ready for v1 

- reply funtionality
- Markdown editor 
- attachments 
- attachments preview
- mentions (tagging users)
- cameracapture , audio recording

Things that will be done in v2 (by Tueday )

makrdown editor (addons to v1) 
- preselect options before we start typing 
- after clearing a word the next words to be plain text instead of previous styles

Mentions
- send a user a notiification that he is tagged (also for reply)
- mention position is not correct when we click on mention button
- filter by name

File: src/Components/Facility/DoctorNoteReplyPreviewCard.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        </div>
+        <div className="max-h-14 overflow-hidden pb-2 pl-14 text-sm text-gray-700">
+          <MarkdownPreview markdown={parentNote.note} />
+        </div>
+      </div>
+      <div>{children}</div>
+    </div>
+  ) : (
+    <div>{children}</div>
+  );
```
Comment (rithviknishad) [inline src/Components/Facility/DoctorNoteReplyPreviewCard.tsx]: A minor code readability improvement

```suggestion
  if (!parentNote) {
    return children;
  }
  
  return (
    <div className="mt-3 flex flex-col rounded-lg border border-gray-300 bg-gray-200 pt-2 text-gray-800">
      <div className="flex flex-col">
        <div className="flex justify-between pl-3">
          <div className="flex gap-2">
            <div className="mt-1 flex h-9 w-9 items-center justify-center rounded-full bg-primary text-lg font-semibold text-white">
              {parentNote.created_by_object?.first_name[0]}
            </div>
            <div>
              <div>
                <span className="text-sm font-semibold text-gray-700">
                  {parentNote.created_by_object?.first_name || "Unknown"}{" "}
                  {parentNote.created_by_object?.last_name}
                </span>
                {parentNote.user_type && (
                  <span className="pl-2 text-sm text-gray-700">
                    {`(${USER_TYPES_MAP[parentNote.user_type]})`}
                  </span>
                )}
              </div>
              <div className="text-xs text-gray-600">
                <div className="tooltip inline">
                  <span className="tooltip-text tooltip-bottom">
                    {formatDateTime(parentNote.created_date)}
                  </span>
                  Created {relativeDate(parentNote.created_date, true)}
                </div>
              </div>
            </div>
          </div>
          {cancelReply && (
            <div
              className="mr-2  cursor-pointer  p-1 text-2xl"
              onClick={cancelReply}
            >
              <CareIcon
                icon="l-times"
                className="m-1 rounded-full  bg-gray-500 p-1 text-white hover:bg-gray-600"
              />
            </div>
          )}
        </div>
        <div className="max-h-14 overflow-hidden pb-2 pl-14 text-sm text-gray-700">
          <MarkdownPreview markdown={parentNote.note} />
        </div>
      </div>
      <div>{children}</div>
    </div>
  );
```

File: src/Components/Facility/PatientNoteCard.tsx [hunk 8]
```diff
@@ -173,7 +282,44 @@ const PatientNoteCard = ({
...
+                onClick={() => {
+                  if (allowThreadView) navigate(`${currentPath}/${note.id}`);
+                }}
+                className={`pl-11 text-sm text-gray-700 ${allowThreadView && "cursor-pointer"}`}
+              >
+                <MarkdownPreview markdown={noteField} />
+                <div className="flex gap-2">
+                  {note.files &&
+                    note.files.length > 0 &&
+                    note.files.map((file: any) => (
```
Comment (rithviknishad) [inline src/Components/Facility/PatientNoteCard.tsx]: 1. A simpler check can be done by using `?` operator right?
2. Any type not allowed. Define proper types for the `files` property.

```suggestion
                  {note.files?.map((file) => (
```

**Review by rithviknishad — CHANGES_REQUESTED** [review]

**Comment by nihal467** [comment]
### TestReport

**Bugs in functionality**

image

- [ ] when you select a text and apply a preference , then remove a preference, and then again give a preference, the selected text will go out of focus

image
image

- [ ] when i select a text preference for example Bold, then start typing text, the text preference is not applied to the text

image

- [ ] when i apply a text preference for a specific text and after applying, continue typing, the text preference is been applied to the new text as well

image

- [ ] when i apply italics and go to next row of the text, now type to un-select italic, it wont allow

image

- [ ] edit and reply buttons are not visible in mobile view

image

- [ ] user tagging will fail if we mention more than 1 character in yours name for example @d if we type it will show the users list but we continue to type @de the list of users wont be visible at all

image

- [ ] when we download the uploaded file , formatting is wrong for example in place of audio.mp3 they are saving as audio.mp3.mp3

**Bug in Design**

image

- [ ] the tagged user text is clickable in a existing note, but when we go and click over it, it's not having any redirection or pop-up or anything relevant ( on last test there was a pop-up showing user name and details currently seem like it is removed)

image

- [ ] the edit and reply button are only visible when we move the cursor near to the place marked as 2 in the above screenshot

image

- [ ] the edit and reply icon is overlapping with notes in mobile view

image

- [ ] when we upload a  profile image, it is still fully reloading the entire page

**Enhancement**

- [ ] try to add multiple file upload support to the platform ( in future wishlist item )

**Review by rithviknishad — APPROVED** [review]
LGTM

**Comment by UdaySagar-Git** [comment]
markdown editor pending 
- preselect/deselect options before we start typing
- after clearing a word the next words to be plain text instead of previous styles

Bugs in functionality
-----
- [ ] when you select a text and apply a preference , then remove a preference, and then again give a preference, the selected text will go out of focus (a video of indicating the issue would be helpful)
- [ x]  when i select a text preference for example Bold, then start typing text, the text preference is not applied to the text (checking this as duplicate- issue 1)
- [ x]  when i apply a text preference for a specific text and after applying, continue typing, the text preference is been applied to the new text as well (thats the expected behaviour - ref : slack editor)
- [ x]  when i apply italics and go to next row of the text, now type to un-select italic, it wont allow  (checking this as duplicate- issue 1)
- [ x]  edit and reply buttons are not visible in mobile view
- [ ]  user tagging will fail if we mention more than 1 character in yours name for example @d if we type it will show the users list but we continue to type @de the list of users wont be visible at all
- [ x]  when we download the uploaded file , formatting is wrong for example in place of audio.mp3 they are saving as audio.mp3.mp3

 
Bug in Design

- [ ] the tagged user text is clickable in a existing note, but when we go and click over it, it's not having any redirection or pop-up or anything relevant ( on last test there was a pop-up showing user name and details currently seem like it is removed) ---yes removed temporaryly to do that from backend
- [x ]  the edit and reply button are only visible when we move the cursor near to the place marked as 2 in the above screenshot
- [ x]  the edit and reply icon is overlapping with notes in mobile view
- [ x]  when we upload a profile image, it is still fully reloading the entire page

**Comment by UdaySagar-Git** [comment]
I will fix the issues at the end of this month. I currently have end-semester exams going on.

Thank you!

**Comment by gigincg** [comment]
@UdaySagar-Git Can you make the changes?

**Comment by UdaySagar-Git** [comment]
@gigincg will fix by end of this week , sorry for the delay.

File: src/Components/Common/MarkdownPreview.tsx (hunk 1 above)
Comment (UdaySagar-Git) [reply src/Components/Common/MarkdownPreview.tsx]: check @UdaySagar-Git

Comment (UdaySagar-Git) [reply src/Components/Common/MarkdownPreview.tsx]: done

**Comment by UdaySagar-Git** [comment]
@rithviknishad @khavinshankar can you review this once

**Comment by bodhish** [comment]
@gigincg can you review this?

**Comment by UdaySagar-Git** [comment]
@nihal467 can you review it once?

https://github.com/user-attachments/assets/193d2775-e83b-417c-af9a-16fdf67c19bd

**Comment by nihal467** [comment]
@UdaySagar-Git can you do the below-mentioned things for review :

- [ ] fix the lint and test in the co-related backend PR
- [ ] The doctor notes related cypress test is failing in the frontend, can you attach the screenshot of the test passing in your local

**Comment by UdaySagar-Git** [comment]
@nihal467 fixed related tests

**Comment by nihal467** [comment]
*Bug*

image

- [ ] the thread view and default view is the doctor notes dedicated page view is breaking UI

image

- [ ] the tooltip of attachment is clipping in the normal full view

*Enhancement*

image

- [ ] the editor will be used by non-tech savy people, for them, when we click bold, rather than "*" , its better to make the text bold itself
- [ ] when the word is getting bold, keep the related icon, as well as in primary colour to show its active

image

- [ ] when drag select multi line para and try to apply order list, it is being only applied to the first line of the para

image

- [ ] add a validation to the hyperlinking related to fields in the pop-up since its accepting empty input

**Comment by UdaySagar-Git** [comment]
> - [ ]  the editor will be used by non-tech savy people, for them, when we click bold, rather than "*" , its better to make the text bold itself
> - [ ]  when the word is getting bold, keep the related icon, as well as in primary colour to show its active

@khavinshankar @bodhish @nihal467 

There is a preview section (clicking on it shows how the markdown will appear in the UI). However, if we want a full-fledged live markdown editor, we need to rely on an external package. Achieving this natively would make the code unmaintainable, especially since the [execCommand()](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand) method, previously used for this, has been deprecated and there is no straightforward alternative.

Here is an implementation of a rich text editor using different libraries: https://github.com/UdaySagar-Git/Text-Editor (Lexical is missing). Please let me know if you're okay with adding a package, and I will make a follow-up PR for this. However, I think relying on plain markdown is also a valid approach (this is how WhatsApp handles it, and users are very familiar with that atleast for just bold and italic).

image


- [x]  the thread view and default view is the doctor notes dedicated page view is breaking UI
- [x]  the tooltip of attachment is clipping in the normal full view
- [x]  when drag select multi line para and try to apply order list, it is being only applied to the first line of the para
- [x]  add a validation to the hyperlinking related to fields in the pop-up since its accepting empty input

**Comment by bodhish** [comment]
@UdaySagar-Git @nihal467 @gigincg can we get this merged

**Comment by nihal467** [comment]
@khavinshankar can you clear the merge conflict and lets get it merged, we will take all other related issues as enhancement

*Conditional approving it*

**Comment by khavinshankar** [comment]
@nihal467 will resolve the conflicts and get it reviewed and merged tomorrow morning

File: src/Components/Common/RichTextEditor/CameraCaptureModal.tsx
Comment (rithviknishad) [inline src/Components/Common/RichTextEditor/CameraCaptureModal.tsx:1]: Not a fan of having camera capture logic duplicated multiple places.

File: src/Components/Common/RichTextEditor/MarkdownPreview.tsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+}: {
+  markdown: string;
+  mentioned_users?: UserBareMinimum[];
+}) => {
+  const MentionedUsers: Record<string, UserBareMinimum> = {};
+  if (mentioned_users) {
+    mentioned_users.forEach((user) => {
+      MentionedUsers[user.username] = user;
+    });
+  }
```
Comment (rithviknishad) [inline src/Components/Common/RichTextEditor/MarkdownPreview.tsx]:
```suggestion
  const MentionedUsers = Object.fromEntries(mentioned_users.map((u) => [u.username, user]))
```

File: src/Components/Common/RichTextEditor/RichTextEditor.tsx [hunk 10]
```diff
@@ -0,0 +1,850 @@
...
+        null,
+        () => {
+          Notification.Error({
+            msg: "Error Uploading File: Network Error",
+          });
+          reject();
+        },
+      );
+    });
+  };
```
Comment (rithviknishad) [inline src/Components/Common/RichTextEditor/RichTextEditor.tsx]: we already have the `uploadFile` method as a shared utility function. Why duplicate it here again?

File: src/Components/Users/ProfilePicUploadModal.tsx
Comment (rithviknishad) [inline src/Components/Users/ProfilePicUploadModal.tsx:1]: This component is not referenced anywhere, let's delete this file.

File: src/Components/Facility/models.tsx [hunk 11]
```diff
@@ -566,6 +566,7 @@ export interface BaseUserModel {
   last_name: string;
   user_type: string;
   last_login: string;
+  read_profile_image_url: string;
```
Comment (rithviknishad) [inline src/Components/Facility/models.tsx]: are you sure the variable name is correct? 

```suggestion
  read_profile_picture_url: string;
```

File: src/Components/Facility/PatientNotesDetailedView.tsx [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+    }
+  }, [reload]);
+
+  useEffect(() => {
+    fetchNotes();
+  }, [thread, noteId]);
+
+  useEffect(() => {
+    setReload?.(true);
+  }, []);
```
Comment (rithviknishad) [inline src/Components/Facility/PatientNotesDetailedView.tsx]: ⚠️ unreadable / hard to understand. comments might help

**Review by rithviknishad — CHANGES_REQUESTED** [review]

File: src/Components/Facility/PatientNotesDetailedView.tsx (hunk 12 above)
Comment (UdaySagar-Git) [reply src/Components/Facility/PatientNotesDetailedView.tsx]: updated with comments

File: src/Components/Common/RichTextEditor/RichTextEditor.tsx (hunk 10 above)
Comment (UdaySagar-Git) [reply src/Components/Common/RichTextEditor/RichTextEditor.tsx]: just upadated to use existing hook (i am not aware of it since it got recently updated)

**Review by rithviknishad — APPROVED** [review]

**Comment by UdaySagar-Git** [comment]
[update] fixed, looks like it was issue with my local git config

@rithviknishad @khavinshankar  git isn't tracking file paths correctly , it is showing differntly in my local and in git 
tried recloning , moving those files completely to new path but no luck, can you help me with this

image

my local path 
image

even changing the location of files not changing path in git

**Review by khavinshankar — APPROVED** [review]

**Comment by nihal467** [comment]
- All the features much as (File Uploads, user linking, etc...) are failing , since its not reviewable and functional closing this PR

**Pull request closed without merging**